pyang/plugins/check_update.py

```diff
@@ -34,6 +34,13 @@ def add_opts(self, optparser):
                                  help=os.pathsep + "-separated search path" \
                                      " for yin and yang modules used by" \
                                      " OLDMODULE"),
+            optparse.make_option("-D", "--check-update-from-deviation-module",
+                                 dest="old_deviation",
+                                 default=[],
+                                 action="append",
+                                 help="Deviation module of the OLDMODULE, or" \
+                                      " module augmenting the OLDMODULE. This" \
```

Comment:
I get the deviation part, but why the augment?  If you want to check a module with augments, just check that module.
(whoops, turns out that didn't work, now fixed in 02d2de4)

Reply:
Sorry for this lengthy comment. Let's say in current folder, b.yang is our base model, and c.yang augments b.yang:

module c {
  namespace urn:c;
  prefix c;

  import b {
    prefix b;
  }

  revision 2014-03-01;

  augment /b:b {
    leaf foo {
      type int32;
    }
  }
}

And d.yang deviates b.yang:

module d {
  namespace urn:d;
  prefix d;

  import b {
    prefix b;
  }

  revision 2014-03-01;

  deviation /b:b/b:q {
    deviate add {
      default 'dummy';
    }
  }
}


In folder ./2014-05-01, b.yang remains same, but both b.yang and c.yang has backward incompatible change:

module c {
  namespace urn:c;
  prefix c;

  import b {
    prefix b;
  }

  revision 2014-05-01;

  augment /b:b {
    leaf foo {
      type int64;
    }
  }
}

module d {
  namespace urn:d;
  prefix d;

  import b {
    prefix b;
  }

  revision 2014-05-01;

  deviation /b:b/b:q {
    deviate add {
      default 'smarty';
    }
  }
}

As you suggested, incompatible changes introduced by augmentations can be checked by:

yuekyang-> pyang --check-update-from c.yang --check-update-from-path . -p 2014-05-01 2014-05-01/c.yang
2014-05-01/c.yang:13: error: the base type has illegally changed from int32 to int64
yuekyang->

Alternatively, it can be checked by:

yuekyang-> pyang --check-update-from b.yang --check-update-from-path . --check-update-from-deviation-module c.yang -14-05-01 2014-05-01/b.yang 2014-05-01/c.yang
2014-05-01/c.yang:13: error: the base type has illegally changed from int32 to int64
yuekyang->

The underlying question is: do we track incompatible changes of model b (at least from customers point of view, they experience behavior changes of model b) in revision of c.yang or b.yang? If the answer is c.yang, the first pyang command line makes more sense. Otherwise, the second payng command would be more reasonable.

A consequence of tracking changes of model b in revision of c.yang is: customers then need to check revision of both b.yang and c.yang to guarantee compatible behavior of model b, and they need to know which modules augments b.yang.

The same question goes to deviation. We can track incompatible changes of model b by revision of b.yang or d.yang. Any guidance from NETMOD group? RFC7950 section 5.1.1 does not mention augment and deviation.

Reply:
module c doesn't change b, so it should be checked on it own.

module d is a deviation module that changes the implementation of b, so it makes sense to check b+d with b'+d' (or b+d', or b'+d).